torchaudio/datasets/dr_vctk.py

```diff
@@ -0,0 +1,116 @@
+from pathlib import Path
+from typing import Dict, Tuple, Union
+
+from torch import Tensor
+from torch.utils.data import Dataset
+
+import torchaudio
+from torchaudio.datasets.utils import (
+    download_url,
+    extract_archive,
+    validate_file,
+)
+
+
+_URL = "https://datashare.ed.ac.uk/bitstream/handle/10283/3038/DR-VCTK.zip"
+_CHECKSUMS = {
+    "https://datashare.ed.ac.uk/bitstream/handle/10283/3038/DR-VCTK.zip": "29e93debeb0e779986542229a81ff29b",
+}
+_SUPPORTED_SUBSETS = {"train", "test"}
+
+
+class DR_VCTK(Dataset):
+    """Create a dataset for Device Recorded VCTK (Small subset version).
+
+    Args:
+        root (str or Path): Root directory where the dataset's top level directory is found.
+        download (bool):
+            Whether to download the dataset if it is not found at root path. (default: ``False``).
+        url (str): The URL to download the dataset from.
+            (default: ``"https://datashare.ed.ac.uk/bitstream/handle/10283/3038/DR-VCTK.zip"``)
+        subset (str): The subset to use. Can be one of ``"train"`` and ``"test"``. (default: ``"train"``).
+    """
+
+    def __init__(
+        self,
+        root: Union[str, Path],
+        download: bool = False,
+        url: str = _URL,
+        subset: str = "train",
+    ) -> None:
+        if subset not in _SUPPORTED_SUBSETS:
+            raise RuntimeError(
+                f"The subset '{subset}' does not match any of the supported subsets: {_SUPPORTED_SUBSETS}"
+            )
+
+        root = Path(root).expanduser()
+        archive = root / "DR-VCTK.zip"
+
+        self._subset = subset
+        self._path = root / "DR-VCTK" / "DR-VCTK"
+        self._clean_audio_dir = self._path / f"clean_{self._subset}set_wav_16k"
+        self._noisy_audio_dir = self._path / f"device-recorded_{self._subset}set_wav_16k"
+        self._config_filepath = self._path / "configurations" / f"{self._subset}_ch_log.txt"
+
+        if not self._path.is_dir():
+            if not archive.is_file():
+                if not download:
+                    raise RuntimeError("Dataset not found. Please use `download=True` to download it.")
+                checksum = _CHECKSUMS.get(url, None)
+                download_url(url, root)
+            with open(archive, "rb") as file_obj:
```

Comment:
The variable `checksum` has to be defined right before performing the check.
In the current way, if the constructor is called when the archive exists but the directory does not, then it will fail.
Also, this will fail if custom URL is provided, as the result of `checksum = _CHECKSUMS.get(url, None)` is `None`.

The URL parameter is useful if the client code has some network restriction, but since we expect the archive to be same even if it is downloaded from somewhere different (say, re-host) can we simply use the same checksum?

If a user really wants to use archive that does not have the same checksum, they can extract it by themselves (but keeping the same directory structure) an the whole logic here is skipped.

Finally, since the logic around it is getting long, can you put this in a helper method?